fplll/pruner/pruner.h

```diff
@@ -290,37 +382,186 @@ template <class FT> class Pruner
       symmetry_factor = 1;
     }
     min_pruning_coefficients.resize(d);
+    btmp.resize(d);
+    bftmp.resize(n);
     fill(min_pruning_coefficients.begin(), min_pruning_coefficients.end(), 0.);
-
     set_tabulated_consts();
+    // need to fix target if possible
+    if (metric == PRUNER_METRIC_PROBABILITY_OF_SHORTEST)
+    {
+      // if target > 1 or < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (this->target > 1.0 || this->target < 0.0)
+      {
+        this->target = 0.99;
+        opt_overall  = true;
+      }
+    }
+    else if (metric == PRUNER_METRIC_EXPECTED_SOLUTIONS)
+    {
+      // if target < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (this->target < 0.0)
+      {
+        this->target = 0.99;
+        opt_overall  = true;
+      }
+    }
+    else
+    {
+      throw std::invalid_argument("Pruner was set to an unknown metric");
+    }
     load_basis_shapes(gso_rs);
   }
 
   /**
+     @brief run the optimization process using 'even' coefficients. Note
+     the optimization only applies to the pruning coefficients indexed
+     by (0, 2, 4, ... n). It thus uses half of the coefficients.
+
      Run the optimization process, successively using the algorithm activated
+     using using half coefficients.
 
-     @brief run the optimization process
+     Note the input pr has length n; but only the even indices in the vector
+     will be used in the optimization. In the end, we have pr_i = pr_{i+1}.
   */
-  void optimize_coefficients(/*io*/ vector<double> &pr);
+  void optimize_coefficients_evec(/*io*/ vector<double> &pr);
+
+  /**
+     @brief run the optimization process using all the coefficients.
+
+     Run the optimization process, successively using the algorithm activated
+     using using full coefficients.
+
+     Note the input pr has length n.
+  */
+  void optimize_coefficients_full(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to reduce single enumeration time
+
+     Optimization process to the pruning parameters to reduce single enumeration
+     time with the hope that it also reduces the overall enumeration time.
+  */
+  void optimize_coefficients_tune_single(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to increase succ. probability
+
+     Optimization process to the pruning parameters to increase succ.
+     probability with the restriction that the single enumeration time does
+     not increase significantly.
+  */
+  void optimize_coefficients_tune_prob(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to make the curve a more smooth.
+
+     Optimization process to the pruning parameters to make the curve
+     more smooth if the input has a large gap between consecutive
+     pruning parameters.
+  */
+  void optimize_coefficients_smooth(/*io*/ vector<double> &pr);
```

Comment:
Same as above, not clear if this is an internal, private helper function or a high level function from the documentation.

Reply:
Changed similarly as mentioned above.